[#119] Add JSON canonical tests from Byron release

Jimbo4350 opened the pull request:
Closes #119

mhuesch commented:
@Jimbo4350 Also, will other JSON instances be brought over from `cardano-sl` to `cardano-chain` later, or does this represent all of them? I made up two cursory lists of types with CanonicalJson instances by grepping in the two repo's - are some of those types missing or do they represent refactoring / renaming where the old types are no longer needed?

[cardano-chain_canonical_json.txt](https://github.com/input-output-hk/cardano-chain/files/2842192/cardano-chain_canonical_json.txt)
[cardano-sl_canonical_json.txt](https://github.com/input-output-hk/cardano-chain/files/2842193/cardano-sl_canonical_json.txt)

Jimbo4350 commented on test/Test/Cardano/Chain/Slotting/Gen.hs:
```diff
@@ -42,6 +44,10 @@ import Cardano.Crypto (ProtocolMagicId)
 
+genCanonicalEpochIndex :: Gen EpochIndex
+genCanonicalEpochIndex =
+  EpochIndex <$> Gen.word64 (Range.constant 0 maxCanonicalLovelaceVal)
```
We are actually removing this. Canonical JSON uses 54-bit integers. My understanding is @ruhatch wrote instances to avoid overflow errors (e.g https://github.com/input-output-hk/cardano-prelude/blob/master/src/Cardano/Prelude/Json/Canonical.hs#L104)

However a couple of the types' Canonical JSON instances are converting them to `Int54`s. In my debugging I restricted `genCanonicalEpochIndex` to only produce indicies within the bounds of `Int54`. We decided instead of having a `genCanonicalEpochIndex`, we would modify the types that have Canonical instances that demand an `Int54`. Is this clear?

mhuesch replied:
> We are actually removing this. Canonical JSON uses 54-bit integers. My understanding is @ruhatch wrote instances to avoid overflow errors (e.g https://github.com/input-output-hk/cardano-prelude/blob/master/src/Cardano/Prelude/Json/Canonical.hs#L104)

Ok, right. Those instances should either wrap the JSNum's into the smaller types (Word32/16, Int32) or promote them (no wrapping since the # of bits is greater than 54) into Word64, Integer, Natural.

> However a couple of the types' Canonical JSON instances are converting them to `Int54`s. In my debugging I restricted `genCanonicalEpochIndex` to only produce indicies within the bounds of `Int54`. We decided instead of having a `genCanonicalEpochIndex`, we would modify the types that have Canonical instances that demand an `Int54`. Is this clear?

I'm a little confused by the wording here - "We decided instead of having a `genCanonicalEpochIndex`" - but we have that, don't we?

Jimbo4350 commented:
> @Jimbo4350 Also, will other JSON instances be brought over from `cardano-sl` to `cardano-chain` later, or does this represent all of them? I made up two cursory lists of types with CanonicalJson instances by grepping in the two repo's - are some of those types missing or do they represent refactoring / renaming where the old types are no longer needed?
> 
> [cardano-chain_canonical_json.txt](https://github.com/input-output-hk/cardano-chain/files/2842192/cardano-chain_canonical_json.txt)
> [cardano-sl_canonical_json.txt](https://github.com/input-output-hk/cardano-chain/files/2842193/cardano-sl_canonical_json.txt)

These are the canonical JSON tests in Byron:
https://github.com/input-output-hk/cardano-sl/blob/ccca8b346413f12e9fad8cab4c855e69669a2151/lib/test/Test/Pos/Genesis/CanonicalSpec.hs

Golden test for `GenesisData`
https://github.com/input-output-hk/cardano-sl/blob/577c82944042ca46a916ba86683b5db149acd49e/chain/test/Test/Pos/Chain/Genesis/Json.hs#L141

I implemented the tests above. The other tests I implemented were mainly for debugging purposes. Most of those types in your grep search are subtypes of `GenesisData` so we should be ok.

mhuesch commented:
@Jimbo4350 Thanks! Sorry for being hard-nosed but wanted to make sure I'm not missing something. Seems like the list of types I see in `cardano-sl` which aren't reflected verbatim in `cardano-chain` is
```
GenesisProtocolConstants
SharedSeed
Timestamp
VssCertificate
VssCertificatesMap
VssMaxTTL
VssMinTTL
ProxySKHeavy
```
I assume those just have different names in `cardano-chain` or have been revised into different types. Looks good to me then 👍

mhuesch commented:
bors r+
